### ISSUE TITLE
Incompatible with C++20 modules in MSVC

### ISSUE THREAD

**Issue by andylizf** [nlohmann/json#3970]
### Description

Importing a module file using the lib will cause a compilation error.

### Reproduction steps

Codes below.

### Expected vs. actual results

To compile successfully.

### Minimal code example

network.ixx:

```c++
module;
#include <nlohmann/json.hpp>
#include <string>
export module network;
using nlohmann::json;
export struct message {
  message(std::string str) { json doc = json::parse(str); }
};
```

main.cpp:
```c++
import network;
int main() { message msg(R"({"hello": "world"})"); }
```

### Error messages

```Shell
nlohmann\json.hpp(117): error C2039: 'json_sax_dom_callback_parser': is not a member of 'nlohmann::json_abi_v3_11_2::detail'
nlohmann\detail\output\serializer.hpp(35): note: see declaration of 'nlohmann::json_abi_v3_11_2::detail'
main.cpp(3): note: see reference to class template instantiation 'nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char>>,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>' being compiled
nlohmann\byte_container_with_subtype.hpp(101): warning C4820: 'nlohmann::json_abi_v3_11_2::byte_container_with_subtype<BinaryType>': '7' bytes padding added after data member 'nlohmann::json_abi_v3_11_2::byte_container_with_subtype<BinaryType>::m_has_subtype'
        with
        [
            BinaryType=std::vector<uint8_t,std::allocator<uint8_t>>
        ]
        [            BinaryType=std::vector<uint8_t,std::allocator<uint8_t>>
        ]
nlohmann\json.hpp(4187): warning C4820: 'nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char>>,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>': '7' bytes padding added after data member 'nlohmann::json_abi_v3_11_2::basic_json<std::map,std::vector,std::basic_string<char,std::char_traits<char>,std::allocator<char>>,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_11_2::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::m_type
```

### Compiler and operating system

Visual Studio 17 2022 MSVC 19.36.32323; Windows-10.0.22624

### Library version

3.11.2

### Validation

- [ ] The bug also occurs if the latest version from the [`develop`](https://github.com/nlohmann/json/tree/develop) branch is used.
- [ ] I can successfully [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests).

**Comment by nlohmann** [comment]
Maybe related to #3974 and fixed by #3975.

**Comment by andylizf** [comment]
Is it fixed now?
BTW, how to use the new version in xmake?

**Comment by gviot** [comment]
it's not fixed, I got this compilation error as well with single include and modules.
It looks kinda weird, not sure why this line in particular errors but not the one above which is almost identical.

**Comment by kiraYuukiAsuna** [comment]
same issue in version 3.11.3

**Comment by rnikander** [comment]
I have the issue, 3.11.3, Visual Studio, using C++20 modules.

**Comment by rioki** [comment]
The issue is not an issue with nlohmann/json, but rather MSVC compiler. I get the same issue with c42f/tinyformat. The compiler does not like something in the header and I just have not yet figured out what it is.

You can work around it if you add a "compilation firewall" by defining an interface in a header and shoving the include in a regular cpp. Too bad both libraries rely heavily on template to make the client code nice and small.

**Comment by JamieSharpe** [comment]
I'm having a similar issue when trying to include `nlohmann/json.hpp` within a C++ module file.

|Dev Attribube|Value|
|-|-|
|Visual Studio Version:|17.10.0 Preview 7.0|
|Windows SDK Version:|10.0|
|Platform Toolset:|Visual Studio 2022 (v143)|
|C++ Language Standard:|Preview - Features from the Latest C++ Working Draft (/std:c++latest)|
|nlohmann-json|Version: 3.11.3|

The following section of my code:

```c++
nlohmann::json config = {
	{"Version", "x.x.x.d"}
};
```

Provides the error:

```
>Main.cpp
> *\vcpkg_installed\x64 - windows\x64 - windows\include\nlohmann\json.hpp(120, 38) : error C2039 : 'json_sax_dom_callback_parser' : is not a member of 'nlohmann::json_abi_v3_11_3::detail'
```

I'm in a greenfield project and would love to be able to use this library to parse and export json data.

@rioki, are you able to expand on your temporary fix until either this library or MSVC have a solution?

Many thanks.

**Comment by rnikander** [comment]
For my project I moved to jsoncpp, which has been working fine with a module wrapper that I made.

**Comment by rioki** [comment]
@JamieSharpe I dropped modules altogether for the time being. I ran into 1001 issues like this. I will revisit modules in a year or two. Looks promising, not there yet.

**Comment by duerrbaby** [comment]
> The issue is not an issue with nlohmann/json, but rather MSVC compiler. I get the same issue with c42f/tinyformat. The compiler does not like something in the header and I just have not yet figured out what it is.
> 
> You can work around it if you add a "compilation firewall" by defining an interface in a header and shoving the include in a regular cpp. Too bad both libraries rely heavily on template to make the client code nice and small.

Exactly how did you implement the compilation firewall? Perhaps you can show some code. I'm getting compilation errors when I try to do it. :/

**Comment by Blechwolf** [comment]
> Exactly how did you implement the compilation firewall? Perhaps you can show some code. I'm getting compilation errors when I try to do it. :/

For me the following works:

JSONTools.ixx

```
module;

#include <filesystem>

export module JSONTools;

import DataDefs;  // defines A

namespace myns
{
    export class JSONTools final
    {
        public:
            static A readA(const std::filesystem::path& filePath);
            static void writeA(const std::filesystem::path& filePath, const A& a);
        private:
            JSONTools() = delete;
            JSONTools(const JSONTools&) = delete;
            JSONTools(JSONTools&&) = delete;
    };    

} // namespace myns
```

DataDefs.cpp

```
module;

#include <fstream>
#include <nlohmann/json.hpp>

module JSONTools;

import DataDefs;

namespace myns
{

    A JSONTools::readA(const std::filesystem::path& filePath);
    {
        std::string fileContent = readTxt(filePath);   // readTxt defined somewhere else
        nlohmann::json jsonContent = nlohmann::json::parse(fileContent);

        A result{};

        A.setVersion(jsonContent["version"].get<int>());

        // Fill A with the data read from the json

        return result;
    }

    void JSONTools::writeA(const std::filesystem::path& filePath, const A& a);
    {
        auto oFile = std::make_unique<std::ofstream>(filePath, std::ios::ate);

        nlohmann::json jsonOutput;

        jsonOutput["version"] = 1;
        // add some more content to json file

        (*oFile) << jsonOutput; 
    }

} // myns
```

CMakeLists.txt

```
project(JSONTools CXX)

find_package(nlohmann_json CONFIG REQUIRED)

set(INTERFACE_LIST
   JSONTools.ixx
)

set(SOURCE_LIST
   DataDefs.cpp
)

add_library(${PROJECT_NAME})

target_sources(${PROJECT_NAME}
   PRIVATE
      ${SOURCE_LIST}
   PUBLIC
   FILE_SET CXX_MODULES
   FILES ${INTERFACE_LIST}
)

target_link_libraries(${PROJECT_NAME}
PRIVATE
   DataDefs
   nlohmann_json::nlohmann_json
)

```

For every other module I am just importing `JSONTools` in the `target_link_libraries` section of the corresponding `CMakeLists.txt`

**Comment by IMFJS** [comment]
We are also running into this issue. Has anyone raised an issue about this with the MSVC team yet? @rioki 
It appears that a Microsoft engineer already ran into this exact issue while investigating another [issue report](https://developercommunity.visualstudio.com/t/c-modules-error-c3200-importing-a-template-with-a/1485765#T-N1486839). Unfortunately, this report seems to have been stale for years.

This issue is the last roadblock for us to finally migrate parts of our codebase to C++ modules.
Dev Attribube | Value
-- | --
Visual Studio Version: | 17.10.7
Windows SDK Version: | 10.0.18362.0
Platform Toolset: | Visual Studio 2022 (v143)
C++ Language Standard: | Preview - Features from the Latest C++ Working Draft (/std:c++latest)
nlohmann-json | Version: 3.11.3

**Comment by saekiamae** [comment]
To fix this, replace in your json.hpp file (around line 20120):

`
    template<typename BasicJsonType, typename InputAdapterType>
    friend class ::nlohmann::detail::json_sax_dom_callback_parser;
`
with
`
    template<typename BasicJsonType, typename InputAdapterType>
    friend class nlohmann::detail::json_sax_dom_callback_parser;
`
(remove :: from between friend class and nlohmann namespace)

**Comment by nlohmann** [comment]
Can anyone confirm this? If so, I'm happy to see a PR for this :)

**Comment by saekiamae** [comment]
All I can say is that i was getting compilation error from this issue and issue mentioned here:
https://github.com/nlohmann/json/issues/4529

and removing these '::' did fix my issue. Not sure if it is universal solution or only solution for my case but seems working to me.

**Comment by ghost** [comment]
Hey guys,
we are using C++20 modules with this library. Most of the stuff works fine
(We are using latest cmake, latest msvc and C++23 for import std).

Basically what we did is putting this inside our nlohman_json module

```

// Global module fragment where #includes can happen
module;

// included headers will not be exported to the client
#include <nlohmann/json.hpp>

// first thing after the Global module fragment must be a module command
export module my_nlohmann_json;

// build up namespace from macros
#define NLOHMANN_NS NLOHMANN_JSON_NAMESPACE_CONCAT( \
  NLOHMANN_JSON_ABI_TAGS, \
  NLOHMANN_JSON_NAMESPACE_VERSION)

export
{
  namespace nlohmann::NLOHMANN_NS::detail
  {
    using ::nlohmann::NLOHMANN_NS::detail::json_sax_dom_callback_parser;
    using ::nlohmann::NLOHMANN_NS::detail::json_sax_dom_parser;
  }

  namespace nlohmann::NLOHMANN_NS::literals
  {
    using ::nlohmann::NLOHMANN_NS::literals::operator""_json;
    using ::nlohmann::NLOHMANN_NS::literals::operator""_json_pointer;
  }



  namespace nlohmann
  {
    using ::nlohmann::basic_json;
    using ::nlohmann::json;
    using ::nlohmann::json_sax;
    using ::nlohmann::ordered_json;
    using ::nlohmann::detail::json_sax_dom_parser;
    using ::nlohmann::detail::json_sax_dom_callback_parser;
    using ::nlohmann::to_json;
    using ::nlohmann::from_json;
  }
}

module : private;
// possible to include headers here, but they will not be exported to the client
```

This works fine for most of our use cases.
However the makros (like `NLOHMANN_DEFINE_TYPE_INTRUSIVE ` cause a major issue).
Since makros are not beeing exported by modules we have to include `#include <nlohmann/json.hpp>`  in our headers which cause stl types from classic includes and modules to collide.

Best case is the avoid name collisions (i.e. don't mix `import std.compat` and `#include <string>` for good.
If you do need to mix stl includes and imports you have to ensure for _**every**_ translation unit that all includes come before the first import.


@nlohmann  do you think it ts possible to extract all makors to a seperate file which ony uses either` import std.compat` or `include <xxx>`

If that is possible the approach suggestion above should work :)

**Comment by gregmarr** [comment]
@Ju1He1 Have you tried including https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/macro_scope.hpp?  I don't know if it will work stand-alone, but it's a possibility.  It should at least give some feedback on what would be needed in a standalone header.

**Comment by ghost** [comment]
Hey @gregmarr 
Thx I allready checked that but [Macro Scope](https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/macro_scope.hpp#L11) `includes <utility>` which breaks `import std` sadley
( i cannot guranrantee for every translation unit that my includes are placed before the imports)

**Comment by IMFJS** [comment]
@Ju1He1 that is an msvc limitation, which apparently is known by @StephanTLavavej and it is being worked on a "long term solution" for that issue (standard library includes coming after import std), although nothing has been done in that regard for almost two years now. It can be worked around by including all standard library headers through a pre-compiled header, thereby ensuring that all their include guards are set (which seems to be the issue with the redefinition errors that would otherwise occur). Not the prettiest workaround, but until Microsoft decides this is worth looking into, it is the best we can do to migrate the C++ ecosystem to modules. Because we as consumers cannot just fork all libraries we depend on and migrate them too, before we can use import std.

**Comment by ghost** [comment]
@Optimierungswerfer 
Did you try it with latest clang + cmake?
Currently we are only using Modules with MSVC. Clang and GCC still use classic includes.
I'm not sure if this is a MSVC only issue or if other compilers have similar issues with the type collisons

**Comment by StephanTLavavej** [comment]
> nothing has been done in that regard for almost two years now.

Sorry - we've been drowning in other work. (I am currently the only maintainer able to spend time on MSVC's STL.) This is still on my priority list of things to do, but it will mostly be compiler work, and the compiler team has also been busy with other tasks.

**Comment by IMFJS** [comment]
@Ju1He1 I did not try this with the other compilers, since we don't use them in our C++23 projects yet. My last knowledge is that msvc had the best module support so far. I can't tell whether the others caught up. This specific issue has been confirmed to be an msvc limitation though at some hard to find places (e.g. https://www.youtube.com/watch?v=Dk_C_E8AtRs&t=1990s or also a small side note in an old MSVC STL release note on GitHub) and unfortunately even the [Microsoft support is confused](https://developercommunity.visualstudio.com/t/Conflicting-instructions-for-mixing-inc/10790461) about this as they told various people (myself included) multiple times that this was intended behavior, which led to even greater confusion.

@StephanTLavavej I genuinely appreciate the clarification on this, thank you. Really looking forward to a solution there, because this has been holding my team back for years now. We optimistically tried to implement C++20 modules in our solution, since [MSVC has been listed as compliant with modules for years](https://en.cppreference.com/w/cpp/compiler_support) but they never seemed to work as advertised, even in 2023 and until this day. Seeing the msvc compiler team working on C++23 and already asking for priorities for C++26, while a major feature of C++20 isn't fully done yet is a bit frustrating.

**Comment by nlohmann** [comment]
FYI: We recently fixed an issue that allows GCC and Clang to use the library as module. See https://github.com/nlohmann/json/tree/develop/tests/module_cpp20 for the CMake project which successfully compiles with the latest GCC and Clang (see https://github.com/nlohmann/json/blob/develop/.github/workflows/ubuntu.yml#L224-L237). I tried to also add a test for MSVC (see https://github.com/nlohmann/json/pull/4768), but failed.

Maybe somebody here can take a look at this @andylizf @Optimierungswerfer @JamieSharpe @rnikander @gviot @StephanTLavavej @saekiamae @duerrbaby @kiraYuukiAsuna

**Comment by ChuanqiXu9** [comment]
> FYI: We recently fixed an issue that allows GCC and Clang to use the library as module. See https://github.com/nlohmann/json/tree/develop/tests/module_cpp20 for the CMake project which successfully compiles with the latest GCC and Clang (see https://github.com/nlohmann/json/blob/develop/.github/workflows/ubuntu.yml#L224-L237). I tried to also add a test for MSVC (see [#4768](https://github.com/nlohmann/json/pull/4768)), but failed.

Hi, I am maintaining https://arewemodulesyet.org/ and I marked nlohmann-json as ready for modules a few days ago incorrectly (I didn't catch it is in test only). And also I feel it is good enough to put it in source code already. And after all, welcome to contribute after nlohmann-json supports modules someday : )

**Comment by nlohmann** [comment]
So the fix for Clang and GCC will be part of the next release (3.12.1 or 3.13.0). For MSVC, I am lacking the experience needed to proceed.

**Comment by StephanTLavavej** [comment]
Can you prepare a self-contained repro? It doesn't need to be minimal, just something that the compiler devs can build and see the error.

**Comment by nlohmann** [comment]
> Can you prepare a self-contained repro? It doesn't need to be minimal, just something that the compiler devs can build and see the error.

What exactly do you need?

**Comment by StephanTLavavej** [comment]
This is the gold standard - a complete source file (or set of source files), the exact command lines used to build them, and the exact compiler output. Here's an example bug report (which is by design because `errno` is a macro, but this is how a user would properly report it if they didn't understand the root cause):

```
C:\Temp>"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvarsall.bat" x64
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.14.0-pre.7.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Temp>type woof.cpp
```
```cpp
#include <assert.h>

#ifdef USE_CLASSIC_INCLUDES
#include <cerrno>
#include <cmath>
#include <cstdio>
#else
import std;
#endif

int main() {
#ifdef USE_CLASSIC_INCLUDES
    std::puts("Classic includes.");
#else
    std::puts("Named modules.");
#endif

    const double dbl{std::sqrt(2.25)};
    const int err{errno};
    std::printf("dbl: %g\n", dbl);
    std::printf("err: %d\n", err);
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od /DUSE_CLASSIC_INCLUDES woof.cpp && woof.exe
woof.cpp
Classic includes.
dbl: 1.5
err: 0

C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od /c "%VCToolsInstallDir%\modules\std.ixx"
std.ixx

C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od woof.cpp std.obj && woof.exe
woof.cpp
woof.cpp(19): error C2065: 'errno': undeclared identifier
```

Once you have this, the proper channel is VS Developer Community, https://developercommunity.visualstudio.com/cpp/report .

It's okay for the set of source files to be large, although obviously smaller/simpler is always better. What's important is making the reproduction steps as simple and bulletproof as possible. "Run these exact command lines" is simple. "Build this zipped IDE solution" is not as simple (there's a lot more complexity and ways for obscure build options to hide in an IDE solution; the compiler devs will still accept it but it's definitely not preferred). "Clone this GitHub repo and follow a bunch of weird steps" is even more complicated and prone to misunderstanding.

(But at least all of those are theoretically self-contained. Bug reports of the form "here's a code *snippet*" or "here's just the error message, no code" are non-actionable and it baffles me when programmers do this to other programmers.)

**Comment by nlohmann** [comment]
Done:

- Repo: https://github.com/nlohmann/msvc_modules_json
- Ticket: https://developercommunity.visualstudio.com/t/nlohmannjson-does-not-compile-as-C20/10908164

**Comment by idingkai** [comment]
> Done:
> 
> * Repo: https://github.com/nlohmann/msvc_modules_json
> * Ticket: https://developercommunity.visualstudio.com/t/nlohmannjson-does-not-compile-as-C20/10908164

Still not working in latest MSVC 17.14.16.
```
error C2039: 'json_sax_dom_callback_parser': is not a member of 'nlohmann::json_abi_v3_12_0::detail'
error C2039: 'unknown_size': is not a member of 'nlohmann::json_abi_v3_12_0::detail'
```

But I eventually found the way to fix it.

json.ixx
```c++
module;

#include <nlohmann/json.hpp>

export module json;

NLOHMANN_JSON_NAMESPACE_BEGIN

namespace detail
{
    export using NLOHMANN_JSON_NAMESPACE::detail::json_sax_dom_callback_parser;
    export using NLOHMANN_JSON_NAMESPACE::detail::unknown_size;
} // namespace detail

export using NLOHMANN_JSON_NAMESPACE::adl_serializer;
export using NLOHMANN_JSON_NAMESPACE::basic_json;
export using NLOHMANN_JSON_NAMESPACE::json;
export using NLOHMANN_JSON_NAMESPACE::json_pointer;
export using NLOHMANN_JSON_NAMESPACE::ordered_json;
export using NLOHMANN_JSON_NAMESPACE::ordered_map;
export using NLOHMANN_JSON_NAMESPACE::to_string;

NLOHMANN_JSON_NAMESPACE_END
```

main.cpp
```c++
#include <iostream>
#include <map>

import json;

int main()
{
    nlohmann::json a = 1;
    nlohmann::json b = 1;
    nlohmann::json c = "string";
    std::cout << "a: " << to_string(a) << std::endl;
    std::cout << (a == b) << std::endl;
    std::cout << (a == c) << std::endl;
}
```
`#include <map>` is necessary, otherwise errors will occur:
```
error C2676: binary '==': 'std::map<StringType,nlohmann::json_abi_v3_12_0::basic_json<std::map,std::vector,StringType,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_12_0::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>,void>,nlohmann::json_abi_v3_12_0::basic_json<std::map,std::vector,StringType,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_12_0::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>,void>::default_object_comparator_t,std::allocator<std::pair<const StringType,nlohmann::json_abi_v3_12_0::basic_json<std::map,std::vector,StringType,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_12_0::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>,void>>>>' does not define this operator or a conversion to a type acceptable to the predefined operator
        with
        [
            StringType=std::basic_string<char,std::char_traits<char>,std::allocator<char>>
        ]
```
or
```
error C2678: binary '<=>': no operator found which takes a left-hand operand of type 'std::map<StringType,nlohmann::json_abi_v3_12_0::basic_json<std::map,std::vector,StringType,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_12_0::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>,void>,nlohmann::json_abi_v3_12_0::basic_json<std::map,std::vector,StringType,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_12_0::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>,void>::default_object_comparator_t,std::allocator<std::pair<const StringType,nlohmann::json_abi_v3_12_0::basic_json<std::map,std::vector,StringType,bool,int64_t,uint64_t,double,std::allocator,nlohmann::json_abi_v3_12_0::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>,void>>>>' (or there is no acceptable conversion)
        with
        [
            StringType=std::basic_string<char,std::char_traits<char>,std::allocator<char>>
        ]
```

**Comment by banderzhm** [comment]
[FIX MVSC](https://github.com/nlohmann/json/pull/4952 ) Look this. Please.

**Comment by nlohmann** [comment]
The fix in https://github.com/nlohmann/json/pull/4952 has a green CI including the test `ci_module_cpp20` running MSVC 19.44.35217.0. Maybe somebody can take a look at the fix and try it. @andylizf @gviot @kiraYuukiAsuna

**Comment by radjkarl** [comment]
cought another issue with msvc 14 and json inside a module:

detail\input\binary_reader.hpp(53): error C2129: static function 'bool nlohmann::json_abi_v3_11_3::detail::little_endianness(int) noexcept' declared but not defined

workaround is to remove `static` from
```
static inline bool little_endianness(int num = 1) noexcept
{
    return *reinterpret_cast<char*>(&num) == 1;
}
```
 
in binary_reader.hpp

**Comment by lyj-514328** [comment]
> Done:
> 
> * Repo: https://github.com/nlohmann/msvc_modules_json
> * Ticket: https://developercommunity.visualstudio.com/t/nlohmannjson-does-not-compile-as-C20/10908164

The MS official response confirms that the issue has been fixed.
